### PR TITLE
Collect tokens from missing SwayParseTree variants

### PR DESCRIPTION
**Pull request by JoshuaBatty**
This PR mostly addresses the missing `Declaration` and `Expression` variants of the `AstNode`s returned by the `SwayParseTree` as mentioned in this issue #1017.

Note, going forward I also plan on using the `TypedParseTree` for collecting the tokens. This is essential for gathering type information about some tokens, but also for accessing the `namespace` associated with each token.

### CODE REVIEW

File: sway-lsp/src/core/token.rs
Comment (mitchmindtree) [inline sway-lsp/src/core/token.rs:426]: Good Q: Is this how the parser represents the `_` syntax? I.e. `let foo: Vec<_> = ...` where `_` explicitly indicates that the type should be inferred? If so, it would be cool if `sway-lsp` could show what type is inferred when you hover over it. That said, that would all have to wait til we introduce type checking anyways.

File: sway-lsp/src/utils/function.rs
Comment (mitchmindtree) [inline sway-lsp/src/utils/function.rs:3]: If we change this to take a span directly instead of a `FunctionDeclaration`, we should add a doc comment to the function that states what assumptions we make about the given `Span`. E.g. it expects a span from either a `FunctionDeclaration`, trait function, or whatever other declarations this is used for. Should probably do the same for `get_function_details` too.

**Review by mitchmindtree — COMMENTED** [review]
Epic, looks like you're getting into the swing of it!

File: sway-core/src/type_engine/engine.rs
Comment (JoshuaBatty) [reply sway-core/src/type_engine/engine.rs:353]: Agh good spot, was experimenting with the `TypedParseTree` and must have missed this.

File: sway-lsp/src/capabilities/hover.rs
Comment (JoshuaBatty) [reply sway-lsp/src/capabilities/hover.rs:70]: Yeah, `extract_visibility` produces a string with trailing space. Which isn't very intuitive, will change this now.

File: sway-lsp/src/capabilities/semantic_tokens.rs
Comment (JoshuaBatty) [reply sway-lsp/src/capabilities/semantic_tokens.rs:81]: I like the idea of having an `enum` with `#[repr(u32)]`. Easy change.

File: sway-lsp/src/core/token.rs
Comment (JoshuaBatty) [reply sway-lsp/src/core/token.rs:426]: Yeah this would be cool, might need to wait for type checking as you say.

File: sway-lsp/src/utils/function.rs
Comment (JoshuaBatty) [reply sway-lsp/src/utils/function.rs:4]: Added comments

**Comment by JoshuaBatty** [comment]
Thanks @mitchmindtree, have addressed each of your comments.

File: sway-lsp/src/capabilities/semantic_tokens.rs
Comment (mitchmindtree) [inline sway-lsp/src/capabilities/semantic_tokens.rs:90]: Rather than including variants for the currently unsupported types, I think you could also do something like:

```rust
#[repr(u32)]
enum TokenTypeIndex {
    Function = 1,
    Namespace = 3,
    Parameter = 5,
    Variable = 9,
    Enum = 10,
    Struct = 11,
    Interface = 12,
}
```

Comment (JoshuaBatty) [reply sway-lsp/src/capabilities/semantic_tokens.rs:90]: Agh nice, didn't know that. Updating now.

**Review by mitchmindtree — APPROVED** [review]
Giving this a utACK tick, but I'll cc @sezna just to check if its OK to make these parse-tree type fields `pub`?

I feel like I remember some talk of potentially removing `sway-core`'s `parse-tree` module in favour of the new `sway-parse` stuff at some point, but even if that's the case I don't know if that should block this as I guess we could just replace this at the same time.

**Review by adlerjohn — COMMENTED** [review]
Looks fine to me, will defer to @sezna

**Comment by sezna** [comment]
>  I remember some talk of potentially removing sway-core's parse-tree module in favour of the new sway-parse stuff at some point

The way things currently work is:
1. `sway-parse` produces a [`Vec<Item>`](https://github.com/FuelLabs/sway/blob/master/sway-parse/src/item/mod.rs#L15) where the items are potentially nested top level things
2. We convert the output of `sway-parse` to the old `ParseTree`
3. The old pipeline continues as normal, unaware of the fact that we didn't use the old parser to generate the parse tree.

Eventually we'd want to skip step 2 above, but it isn't a priority since it works. It'll probably get prioritized after all major features are stabilized, then we can start cleaning up stale/non-optimal code etc.


It is totally fine to make these parse tree items public, as this PR does -- `sway-core`'s external API is more or less just whatever the tooling needs. @JoshuaBatty should look into using `sway-parse::Item` instead, though, as it would be faster and the items are a bit more well-formed.

**Review by sezna — APPROVED** [review]
LGTM but see above comment about potentially switching the lsp over to `sway-parse`.

**Comment by JoshuaBatty** [comment]
Awesome, thanks for that @sezna! I hadn't looked into `sway-parse::Item` yet. Appreciate the feedback, will look at integrating this into `sway-lsp`. Thanks :)

**Review by mitchmindtree — APPROVED** [review]
Re-approving after successful rebase :+1:

**Pull request merged**
